### PR TITLE
Optimized aFreeRDP for Samsung DeX

### PR DESCRIPTION
**Pull request by aleksei-klv**
Adapted Android FreeRDP client for compatibility with Samsung DeX according the guide: http://developer.samsung.com/samsung-dex

Enabled Multi-window feature for Desktop mode
Added Immersive mode
Enabled full mouse support (scroll and right click)
Fixed bugs in Desktop mode

### CODE REVIEW

**Comment by aleksei-klv** [comment]
@bmiklautz, new pull request instead of previous one https://github.com/FreeRDP/FreeRDP/pull/4128. Merge was removed

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by aleksei-klv** [comment]
@bmiklautz could you check my updates? thx

**Comment by aleksei-klv** [comment]
@akallabeth could you review the request?

**Comment by akallabeth** [comment]
@aleksei-klv Your pull looks a lot better now, I just don't like the memory leak. If the app is crashing that means there is still a sync issue buried somewhere.

**Comment by akallabeth** [comment]
@aleksei-klv Just a hinch, you write manually close window, does that mean the RDP session is still running? In that case a call to ```LibFreeRDP.disconnect()``` is required, otherwise you free native resources that are sitll in use.

**Comment by aleksei-klv** [comment]
@akallabeth App crashes in phone multi-window mode too. Bug is not connected with pull request. I opened bug report for crashing issue: https://github.com/FreeRDP/FreeRDP/issues/4169

**Comment by akallabeth** [comment]
@aleksei-klv Yes, that is fine. Just the hacks you have to get it running (the memory leak and DexFixCloseWin) are in this pull. Both are hiding an issue and therefore must be eliminated.
As already mentioned, try to check if the session is still running.
If it is, call  disconnect and wait for ```OnDisconnecting```.
(Native code is calling ```OnConnectionSuccess``` when successfully established and ```OnDisconnecting``` when a running session ends)
 The your crashes should be gone, if my intuition is right.

File: client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/application/GlobalApp.java [hunk 1]
```diff
@@ -106,7 +106,7 @@ static public SessionState getSession(long instance) {
     static public void freeSession(long instance) {
         if (GlobalApp.sessionMap.containsKey(instance)) {
             GlobalApp.sessionMap.remove(instance);
-            LibFreeRDP.freeInstance(instance);
+            //LibFreeRDP.freeInstance(instance);
```
Comment (akallabeth) [inline client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/application/GlobalApp.java]: Memory leak, revert.

File: client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionActivity.java [hunk 2]
```diff
@@ -351,8 +358,12 @@ protected void onDestroy() {
         // remove clipboard listener
         mClipboardManager.removeClipboardboardChangedListener(this);
 
+        // FIXME: prevent crashing app in DeX Mode: app crashes with popup window when user manually closes the Window
+        DexFixCloseWin();
```
Comment (akallabeth) [inline client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/presentation/SessionActivity.java]: Hack (along with the memory leak mentioned above)
From the description this is caused by not stopping the running session and then freeing up the resources still in use by the native code. Remove

**Comment by akallabeth** [comment]
@aleksei-klv You may want to take a look at #4179, if my assumption is correct your multiwindow issues should be resolved with that pull.

**Comment by aleksei-klv** [comment]
@akallabeth Checked in DeX and Multi-Window mode. Nothing changed. App still crashes

**Comment by akallabeth** [comment]
@aleksei-klv Ok, could you compile debug, enable logging and check logcat for the crash reason? As for this pull, I'd merge it when the 2 issues pointed out are fixe (aka removed).

**Comment by aleksei-klv** [comment]
Hello, Team!
I have checked your patch #4179 and enabled logging. Here is logcat output error:

`10-16 14:52:50.998 3849-5464/? E/ActivityTrigger: activityResumeTrigger: not whiteListedcom.sec.android.app.desktoplauncher/com.android.launcher3.Launcher/235401000
10-16 14:52:51.007 3849-4155/? E/SemDvfsHintManager:  mHintTimeout : 800
10-16 14:52:51.014 3849-4155/? E/ActivityTrigger: activityResumeTrigger: not whiteListedcom.sec.android.app.desktoplauncher/com.android.launcher3.Launcher/235401000
10-16 14:52:51.038 2654-2729/com.freerdp.afreerdp E/com.freerdp.client.android: [pid=2654:tid=307ff450] - Failed to check FreeRDP file descriptor
10-16 14:52:51.048 2654-2654/com.freerdp.afreerdp E/AndroidRuntime: FATAL EXCEPTION: main
                                                                    Process: com.freerdp.afreerdp, PID: 2654
                                                                    java.lang.RuntimeException: Unable to destroy activity {com.freerdp.afreerdp/com.freerdp.freerdpcore.presentation.SessionActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Boolean.booleanValue()' on a null object reference
                                                                        at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4670)
                                                                        at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4697)
                                                                        at android.app.ActivityThread.-wrap7(ActivityThread.java)
                                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1720)
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                        at android.os.Looper.loop(Looper.java:154)
                                                                        at android.app.ActivityThread.main(ActivityThread.java:6823)
                                                                        at java.lang.reflect.Method.invoke(Native Method)
                                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1557)
                                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1445)
                                                                     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Boolean.booleanValue()' on a null object reference
                                                                        at com.freerdp.freerdpcore.services.LibFreeRDP.freeInstance(LibFreeRDP.java:102)
                                                                        at com.freerdp.freerdpcore.application.GlobalApp.freeSession(GlobalApp.java:111)
                                                                        at com.freerdp.freerdpcore.presentation.SessionActivity.onDestroy(SessionActivity.java:372)
                                                                        at android.app.Activity.performDestroy(Activity.java:7252)
                                                                        at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1161)
                                                                        at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:4657)
                                                                        at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:4697) 
                                                                        at android.app.ActivityThread.-wrap7(ActivityThread.java) 
                                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1720) 
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                        at android.os.Looper.loop(Looper.java:154) 
                                                                        at android.app.ActivityThread.main(ActivityThread.java:6823) 
                                                                        at java.lang.reflect.Method.invoke(Native Method) 
                                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1557) 
                                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1445) 
10-16 14:52:52.634 3849-5187/? E/KioskModeService: There's no matched data
10-16 14:52:52.668 3849-3916/? E/ViewRootImpl: sendUserActionEvent() returned.
10-16 14:52:54.124 7449-24836/? E/SPPClientService: [b] SharedConstants.WHAT_TIMEOUT
10-16 14:52:55.646 31835-31835/? E/SignalClusterView: setSlotFocusVisible : subId -2
`

**Comment by aleksei-klv** [comment]
If i change line in LibFreeRDP.java
from
` while(mInstanceState.get(inst)) {`
to
` while(mInstanceState.get(inst) == Boolean.TRUE) {`

i'm getting the error below:

> 09-19 18:33:21.575 19448-19448/? A/DEBUG: signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x90
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x0   0000000000000000  x1   0000007e1103e500  x2   0000000000000000  x3   0000000000000020
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x4   0000000000000004  x5   0000007e0b290b90  x6   0000007ddd5ff310  x7   0000000000000000
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x8   0000007ddd5ff308  x9   0000007dfd258cc8  x10  0000000000000004  x11  0000000000000000
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x12  0000000000000001  x13  0000000000000005  x14  00000000000000e0  x15  00000000ffffffff
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x16  0000007dfcf7eb38  x17  0000007e1e33efc0  x18  00000000ffffffff  x19  0000007ddd5ff450
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x20  0000007ddd5ff4d0  x21  000000000000445c  x22  0000007e1102ae00  x23  da1a3eaa968927fd
09-19 18:33:21.575 19448-19448/? A/DEBUG:     x24  0000007ddd5ff4d0  x25  00000000000fd000  x26  da1a3eaa968927fd  x27  0000007e11283508
09-19 18:33:21.576 19448-19448/? A/DEBUG:     x28  0000007dde1fbff0  x29  0000007ddd5ff2c0  x30  0000007dfce83f3c
09-19 18:33:21.576 19448-19448/? A/DEBUG:     sp   0000007ddd5ff2c0  pc   0000007dfce83f7c  pstate 0000000060000000
09-19 18:33:21.582 19448-19448/? A/DEBUG: backtrace:
09-19 18:33:21.582 19448-19448/? A/DEBUG:     #00 pc 000000000007af7c  /data/app/com.freerdp.afreerdp-2/lib/arm64/libfreerdp2.so
09-19 18:33:21.582 19448-19448/? A/DEBUG:     #01 pc 0000000000078668  /data/app/com.freerdp.afreerdp-2/lib/arm64/libfreerdp2.so (freerdp_disconnect+120)
09-19 18:33:21.582 19448-19448/? A/DEBUG:     #02 pc 00000000000049cc  /data/app/com.freerdp.afreerdp-2/lib/arm64/libfreerdp-android2.so
09-19 18:33:21.582 19448-19448/? A/DEBUG:     #03 pc 000000000006f774  /data/app/com.freerdp.afreerdp-2/lib/arm64/libwinpr2.so
09-19 18:33:21.582 19448-19448/? A/DEBUG:     #04 pc 000000000006c4a0  /system/lib64/libc.so (_ZL15__pthread_startPv+208)
09-19 18:33:21.582 19448-19448/? A/DEBUG:     #05 pc 000000000001e220  /system/lib64/libc.so (__start_thread+16)
09-19 18:33:21.970 3771-4296/system_process E/Watchdog: !@Sync 14 [2017-09-19 18:33:21.969]
09-19 18:33:23.121 3316-3316/? E/audit: type=1701 audit(1505835203.114:612): auid=4294967295 uid=10223 gid=10223 ses=4294967295 subj=u:r:untrusted_app:s0:c512,c768 pid=18581 comm="Thread-10" exe="/system/bin/app_process64" sig=11

**Comment by aleksei-klv** [comment]
@akallabeth 
could you check the logs we posted above?

**Comment by akallabeth** [comment]
@aleksei-klv I've already updated the pull accordingly, did the message get lost? Could you retry with the updated version?

**Comment by aleksei-klv** [comment]
@akallabeth The same result. Still doesn't work

**Comment by akallabeth** [comment]
@aleksei-klv Well, then at least the logs must have changed. could you attach them to #4179?

**Comment by aleksei-klv** [comment]
@akallabeth  done

**Comment by akallabeth** [comment]
@freerdp-bot test

**Pull request merged**